### DISCUSSION TITLE
Prism.Ioc Container

### DISCUSSION THREAD

**Discussion by Midnayt** [PrismLibrary/Prism#2452] [Xamarin.Forms]
@brianlagunas Is it possible to cleare registered singltons? f.e. I have a few servers with different data wich can be stored at interfaces and when logout I need to cleare that, because when I'll choose another server some info will be restored from previous server.

**Comment by brianlagunas** [comment]
well, usually you wouldn't need to "clear" a registration. Instead you would update the instance to point to the correct server. Or you could replace the instance with another instance.

**Reply by xusan** [reply]
Hi @brianlagunas 
 How to Replace instance ?

**Comment by sylvesp** [comment]
My understanding is that there is no general way to replace instances.
You have to write code specific to your IoC container.
In my Prism app I am using DryIoC and the code to replace instances is:

```
//*** DryIoc specific ***
   ContainerLocator.Container.GetContainer().RegisterInstance(typeof(IBIMrxApplicationSettings), _commonSettingsEx, IfAlreadyRegistered.Replace);
//*** End DryIoc specific ***

```

**Reply by dansiegel** [reply]
extremely unnecessary as we already replace singletons by default.
https://github.com/PrismLibrary/Prism/blob/839fda8b3be5412f075890b091a5092aca963d4f/src/Containers/Prism.DryIoc.Shared/DryIocContainerExtension.cs#L117 
so all you need to do is `container.RegiserSingleton<ISomeService, SomeOtherImplementation>()` keeping in mind this will update it with the container but anything that had resolved the service already will still have that same instance it won't "magically" update itself.

**Reply by sylvesp** [reply]
Ah! Sorry. I missed a key word in there, and that is **"Singleton"**
In my scenario I have to replace non-singleton registrations.
E.g:
- For a given interface there is a default implementation registered on startup
- The user chooses one of the multiple possible enhancements => We replace the registration of the interface with the enhnced implementation the user did choose.

As long as I know Prism doesn't allow the replacement of non singleton registrations - but as said earlier I may be wrong on this.
For now I am using container specific code to replace non-singleton registrations.

**Reply by brianlagunas** [reply]
Prism has always allowed you replace any registration. This actually isn’t even a Prism thing. It’s a container thing. The replacement behavior depends on the container you’re using.

**Reply by sylvesp** [reply]
:) so I am not going mad after all 
What Brian said is exactly what I've posted in my original reply: If you want to replace non-singleton registration then you have to use container specific code.

```
//*** DryIoc specific ***
   ContainerLocator.Container.GetContainer().RegisterInstance(typeof(IBIMrxApplicationSettings), _commonSettingsEx, IfAlreadyRegistered.Replace);
//*** End DryIoc specific ***
```

Wish everybody a nice weekend.

**Reply by brianlagunas** [reply]
You are going mad 🤣. You don’t need to use container specific code. Look at the code linked above and scroll down to see the standard register method. It uses replace 😁

**Reply by sylvesp** [reply]
Will do it tomorrow. Promise! it is too late for such a delicate thing! 
:D
